Fix empty target_config in apply_rust_config bootstrap

bit-aloo opened the pull request:
This PR fixes the issue of an empty target_config in apply_rust_config, which was caused by the ordering of TOML config parsing. This was inadvertently introduced during the last config refactor. The test and the corresponding configuration order have been corrected in this PR.

r? @Kobzol

bit-aloo committed:
populate the target_config now before its usage and update the test accordingly

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16368233521/job/46250410023) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46250410023) [(plain)](https://github.com/rust-lang/rust/commit/4e53043395b747b267bee975080896fb781fe19f/checks/46250410023/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: doc_core_no_std_target
Source: src/bootstrap/src/core/builder/tests.rs:1634
────────────────────────────────────────────────────────────────────────────────
Expression: ctx.config("doc").path("core").override_target_no_std(&host_target()).render_steps()
────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────
    0     0 │ [build] llvm <host>
    1     1 │ [build] rustc 0 <host> -> rustc 1 <host>
    2       │-[build] rustc 0 <host> -> LldWrapper 1 <host>
    3     2 │ [build] rustdoc 0 <host>
    4     3 │ [doc] std 1 <host> crates=[core]
────────────┴───────────────────────────────────────────────────────────────────
Stopped on the first failure. Run `cargo insta test` to run all snapshots.

thread 'core::builder::tests::snapshot::doc_core_no_std_target' panicked at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/insta-1.43.1/src/runtime.rs:679:13:
snapshot assertion for 'doc_core_no_std_target' failed in line 1634

---- core::builder::tests::snapshot::doc_library_no_std_target stdout ----
Running test in /tmp/.tmpyD8H7p
##[group]Building LLVM for aarch64-unknown-linux-gnu
##[endgroup]
---
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: doc_library_no_std_target
Source: src/bootstrap/src/core/builder/tests.rs:1650
────────────────────────────────────────────────────────────────────────────────
Expression: ctx.config("doc").path("library").override_target_no_std(&host_target()).render_steps()
────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────
    0     0 │ [build] llvm <host>
    1     1 │ [build] rustc 0 <host> -> rustc 1 <host>
    2       │-[build] rustc 0 <host> -> LldWrapper 1 <host>
    3     2 │ [build] rustdoc 0 <host>
    4     3 │ [doc] std 1 <host> crates=[alloc,core]
────────────┴───────────────────────────────────────────────────────────────────
Stopped on the first failure. Run `cargo insta test` to run all snapshots.

thread 'core::builder::tests::snapshot::doc_library_no_std_target' panicked at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/insta-1.43.1/src/runtime.rs:679:13:
snapshot assertion for 'doc_library_no_std_target' failed in line 1650


failures:
    core::builder::tests::snapshot::doc_core_no_std_target
    core::builder::tests::snapshot::doc_library_no_std_target
```

</details>

bit-aloo committed:
add default_lld_opt_in_targets method and modified test accordingly

add test for lld opt in and also add thread_local defined state to change opt in targets

make the config lld test parameter smoother to work, and have no_llvm_config set even when target_config is not present

Kobzol commented on src/bootstrap/src/core/config/toml/rust.rs:
```diff
@@ -409,6 +409,14 @@ pub(crate) fn validate_codegen_backends(backends: Vec<String>, section: &str) ->
+    &["x86_64-unknown-linux-gnu"]
+}
+#[cfg(test)]
+fn default_lld_opt_in_targets() -> &'static [&'static str] {
```
Well ideally this should work also when running in parallel, hence why I suggested a thread local instead of a global variable :)

Kobzol commented on src/bootstrap/src/core/builder/tests.rs:
```diff
@@ -1644,10 +1644,31 @@ mod snapshot {
 
+    #[test]
+    fn test_lld_opt_in() {
+        let target: &'static str = Box::leak(Box::new(host_target()));
```
I don't think we need this kind of leak magic here. Let's just store `Vec<String>` in the thread loca, and return `&[String]` from `default_lld_opt_in_targets`.

Kobzol commented on src/bootstrap/src/core/builder/tests.rs:
```diff
@@ -1644,10 +1644,31 @@ mod snapshot {
+            let ctx = TestCtx::new();
+
+            insta::assert_snapshot!(
+                ctx.config("doc")
```
Let's do `build compiler` here, and get rid of `override_target_no_std`. We don't need to check `doc` here, we just want to see that `LldWrapper` was built.

You can also call `create_config()` on the config builder and assert that LLD was indeed enabled.

Kobzol commented on src/bootstrap/src/core/config/toml/rust.rs:
```diff
@@ -413,9 +413,25 @@ pub(crate) fn validate_codegen_backends(backends: Vec<String>, section: &str) ->
+
+#[cfg(test)]
+thread_local! {
+    static TEST_LLD_OPT_IN_TARGETS: std::cell::RefCell<Option<&'static [&'static str]>> = std::cell::RefCell::new(None);
```
This could just be a `Vec` initialized to an empty vector.

Kobzol reviewed (commented):
Looks good! Left some suggestions, but the general shape is solid.

bit-aloo committed:
add test_lld_opt_in test

Kobzol commented:
@bors r+
